bricks/stm32/main.c

```diff
@@ -189,6 +332,64 @@ mp_obj_t mp_builtin_open(size_t n_args, const mp_obj_t *args, mp_map_t *kwargs)
 }
 MP_DEFINE_CONST_FUN_OBJ_KW(mp_builtin_open_obj, 1, mp_builtin_open);
 
+static mp_obj_t pb_import_loaded_module(size_t n_args, const mp_obj_t *args) {
+
+    // For backwards compatibility, skip imports for legacy single-script data.
+    // TODO: This can be removed once all IDE tools are updated.
+    if (((uint8_t *)mpy_first)[0] == 'M') {
+        return MP_OBJ_NULL;
+    }
+
+    // Try to find the requested module.
+    mpy_info_t *info = mpy_data_find(mp_obj_str_get_str(args[0]));
+    if (!info) {
+        // Not found, so give up.
+        return MP_OBJ_NULL;
+    }
+
+    // Parse the static script data.
+    mp_reader_t reader;
+    pb_reader_user_data_t data;
+    pb_reader_user_data_new(&reader, &data, mpy_data_get_buf(info), info->mpy_size);
+
+    // Create new module and execute in its own context.
+    mp_obj_t module_obj = mp_obj_new_module(mp_obj_str_get_qstr(args[0]));
+    mp_module_context_t *context = MP_OBJ_TO_PTR(module_obj);
+    mp_compiled_module_t compiled_module = mp_raw_code_load(&reader, context);
+    do_execute_raw_code(context, compiled_module.rc, compiled_module.context);
+
+    // Return the module we found.
+    return module_obj;
+}
+
+// Overrides MicroPython's mp_builtin___import__
+mp_obj_t pb_builtin_import(size_t n_args, const mp_obj_t *args) {
+    // Check that it's not a relative import
+    if (n_args >= 5 && MP_OBJ_SMALL_INT_VALUE(args[4]) != 0) {
+        mp_raise_NotImplementedError(MP_ERROR_TEXT("relative import"));
+    }
+
+    // Check if module already exists, and return it if it does
+    qstr module_name_qstr = mp_obj_str_get_qstr(args[0]);
+    mp_obj_t module_obj = mp_module_get_loaded_or_builtin(module_name_qstr);
+    if (module_obj != MP_OBJ_NULL) {
+        return module_obj;
+    }
+
+    // Pybricks extension: Search for modules in downloaded user data.
+    module_obj = pb_import_loaded_module(n_args, args);
+    if (module_obj != MP_OBJ_NULL) {
+        return module_obj;
+    }
+
+    // Couldn't find the module, so fail
+    #if MICROPY_ERROR_REPORTING <= MICROPY_ERROR_REPORTING_TERSE
+    mp_raise_msg(&mp_type_ImportError, MP_ERROR_TEXT("module not found"));
```

Comment:
If we really want this to be terse, we could omit the message altogether. As it is, this message is only a few bytes shorter than the other one, so its not really saving much.

Reply:
This was coming from the default import, but I agree we can drop it in the terse case.